File: DataFeedViews/DistributionValueDataFeedView.vb
'************************************************************************************
' ST-Sim: A .NET class library for developing state-and-transition simulation models.
'
' Copyright © 2007-2017 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.
'
'************************************************************************************

Imports System.Text
Imports System.Reflection
Imports System.Windows.Forms
Imports System.Globalization
Imports SyncroSim.Core
Imports SyncroSim.Core.Forms

<ObfuscationAttribute(Exclude:=True, ApplyToMembers:=False)>
Class DistributionValueDataFeedView

    Private m_Grid As BaseDataGridView
    Private m_TypeId As Nullable(Of Integer)

    Protected Overrides Sub InitializeView()

        MyBase.InitializeView()

        Dim v As DataFeedView = Me.Session.CreateMultiRowDataFeedView(Me.Scenario, Me.ControllingScenario)
        Me.PanelMain.Controls.Add(v)

        Dim ds As DataSheet = Me.Project.GetDataSheet(DISTRIBUTION_TYPE_DATASHEET_NAME)

        Try
            Me.m_TypeId = ds.ValidationTable.GetValue(DISTRIBUTION_TYPE_NAME_UNIFORM_INTEGER)
        Catch ex As Exception
            Debug.Assert(False)
        End Try

    End Sub

    Protected Overrides Sub Dispose(ByVal disposing As Boolean)

        If (disposing And Not Me.IsDisposed) Then

            If (Me.m_Grid IsNot Nothing) Then

                RemoveHandler Me.m_Grid.CellBeginEdit, AddressOf OnGridCellBeginEdit
                RemoveHandler Me.m_Grid.CellEndEdit, AddressOf OnGridCellEndEdit
                RemoveHandler Me.m_Grid.CellFormatting, AddressOf OnGridCellFormatting
                RemoveHandler Me.m_Grid.CellValueChanged, AddressOf OnGridCellValueChanged

            End If

            If components IsNot Nothing Then
                components.Dispose()
            End If

        End If

        MyBase.Dispose(disposing)

    End Sub

    Public Overrides Sub LoadDataFeed(dataFeed As Core.DataFeed)

        MyBase.LoadDataFeed(dataFeed)

        Dim v As MultiRowDataFeedView = CType(Me.PanelMain.Controls(0), MultiRowDataFeedView)
        Me.m_Grid = v.GridControl()
        v.LoadDataFeed(dataFeed, DISTRIBUTION_VALUE_DATASHEET_NAME)

        AddHandler Me.m_Grid.CellBeginEdit, AddressOf OnGridCellBeginEdit
        AddHandler Me.m_Grid.CellEndEdit, AddressOf OnGridCellEndEdit
        AddHandler Me.m_Grid.CellFormatting, AddressOf OnGridCellFormatting
        AddHandler Me.m_Grid.CellValueChanged, AddressOf OnGridCellValueChanged

    End Sub

    Public Overrides Sub EnableView(enable As Boolean)

        If (Me.PanelMain.Controls.Count > 0) Then

            Dim v As DataFeedView = CType(Me.PanelMain.Controls(0), DataFeedView)
            v.EnableView(enable)

        End If

    End Sub

    Private Sub OnGridCellBeginEdit(ByVal sender As Object, ByVal e As DataGridViewCellCancelEventArgs)

        If (e.ColumnIndex = Me.m_Grid.Columns(DISTRIBUTION_VALUE_DIST_TYPE_ID_COLUMN_NAME).Index) Then

            Dim dgv As DataGridViewRow = Me.m_Grid.Rows(e.RowIndex)
            Dim ds As DataSheet = Me.Project.GetDataSheet(DISTRIBUTION_TYPE_DATASHEET_NAME)
            Dim filter As String = CreateUserDistributionTypeFilter(Me.Project)
            Dim dv As New DataView(ds.GetData(), filter, ds.DisplayMember, DataViewRowState.CurrentRows)
            Dim Cell As DataGridViewComboBoxCell = CType(dgv.Cells(DISTRIBUTION_VALUE_DIST_TYPE_ID_COLUMN_NAME), DataGridViewComboBoxCell)

            Cell.DataSource = dv
            Cell.ValueMember = ds.ValueMember
            Cell.DisplayMember = ds.DisplayMember

        ElseIf (e.ColumnIndex = Me.m_Grid.Columns(DISTRIBUTION_VALUE_VALUE_DIST_TYPE_ID_COLUMN_NAME).Index) Then

            Dim dgv As DataGridViewRow = Me.m_Grid.Rows(e.RowIndex)
            Dim ds As DataSheet = Me.Project.GetDataSheet(DISTRIBUTION_TYPE_DATASHEET_NAME)
            Dim filter As String = CreateKnownDistributionTypeFilter(Me.Project)
            Dim dv As New DataView(ds.GetData(), filter, ds.DisplayMember, DataViewRowState.CurrentRows)
            Dim Cell As DataGridViewComboBoxCell = CType(dgv.Cells(DISTRIBUTION_VALUE_VALUE_DIST_TYPE_ID_COLUMN_NAME), DataGridViewComboBoxCell)

            Cell.DataSource = dv
            Cell.ValueMember = ds.ValueMember
            Cell.DisplayMember = ds.DisplayMember

        End If

    End Sub

    Private Sub OnGridCellEndEdit(ByVal sender As Object, ByVal e As DataGridViewCellEventArgs)

        If (e.ColumnIndex = Me.m_Grid.Columns(DISTRIBUTION_VALUE_DIST_TYPE_ID_COLUMN_NAME).Index) Then

            Dim dgv As DataGridViewRow = Me.m_Grid.Rows(e.RowIndex)
            Dim Cell As DataGridViewComboBoxCell = CType(dgv.Cells(DISTRIBUTION_VALUE_DIST_TYPE_ID_COLUMN_NAME), DataGridViewComboBoxCell)
            Dim Column As DataGridViewComboBoxColumn = CType(Me.m_Grid.Columns(DISTRIBUTION_VALUE_DIST_TYPE_ID_COLUMN_NAME), DataGridViewComboBoxColumn)

            Cell.DataSource = Column.DataSource
            Cell.ValueMember = Column.ValueMember
            Cell.DisplayMember = Column.DisplayMember

        ElseIf (e.ColumnIndex = Me.m_Grid.Columns(DISTRIBUTION_VALUE_VALUE_DIST_TYPE_ID_COLUMN_NAME).Index) Then

            Dim dgv As DataGridViewRow = Me.m_Grid.Rows(e.RowIndex)
            Dim Cell As DataGridViewComboBoxCell = CType(dgv.Cells(DISTRIBUTION_VALUE_DIST_TYPE_ID_COLUMN_NAME), DataGridViewComboBoxCell)
            Dim Column As DataGridViewComboBoxColumn = CType(Me.m_Grid.Columns(DISTRIBUTION_VALUE_VALUE_DIST_TYPE_ID_COLUMN_NAME), DataGridViewComboBoxColumn)

            Cell.DataSource = Column.DataSource
            Cell.ValueMember = Column.ValueMember
            Cell.DisplayMember = Column.DisplayMember

        End If

    End Sub

    Private Sub OnGridCellFormatting(ByVal sender As Object, ByVal e As DataGridViewCellFormattingEventArgs)

        If (e.RowIndex < 0) Then
            Return
        End If

        'Is it an index we want?

        If (e.ColumnIndex <> Me.m_Grid.Columns(DISTRIBUTION_VALUE_VALUE_DIST_MIN_COLUMN_NAME).Index And
            e.ColumnIndex <> Me.m_Grid.Columns(DISTRIBUTION_VALUE_VALUE_DIST_MAX_COLUMN_NAME).Index) Then

            Return

        End If

        'Set the default
        e.CellStyle.Format = "0.0000"

        'Does the cell have a value?

        Dim TargetCell As DataGridViewCell = Me.m_Grid.Rows(e.RowIndex).Cells(e.ColumnIndex)

        If (TargetCell.Value Is Nothing Or TargetCell.Value Is DBNull.Value) Then
            Return
        End If

        'Is the distribution type 'Uniform Integer'?

        Dim TypeCell As DataGridViewCell = Me.m_Grid.Rows(e.RowIndex).Cells(DISTRIBUTION_VALUE_VALUE_DIST_TYPE_ID_COLUMN_NAME)

        If (TypeCell.Value Is Nothing Or TypeCell.Value Is DBNull.Value Or (Not Me.m_TypeId.HasValue)) Then
            Return
        End If

        Dim ValTypeId As Integer = CInt(TypeCell.Value)

        If (ValTypeId <> Me.m_TypeId.Value) Then
            Return
        End If

        e.CellStyle.Format = "0"

    End Sub

    Private Sub OnGridCellValueChanged(ByVal sender As Object, ByVal e As DataGridViewCellEventArgs)

        If (e.ColumnIndex = Me.m_Grid.Columns(DISTRIBUTION_VALUE_VALUE_DIST_TYPE_ID_COLUMN_NAME).Index) Then
            Me.m_Grid.InvalidateRow(e.RowIndex)
        End If

    End Sub


    Private Shared Function CreateKnownDistributionTypeFilter(ByVal project As Project) As String

        Dim ids As List(Of Integer) = GetInternalDistributionTypeIds(project)
        Debug.Assert(ids.Count > 0)

        If (ids.Count = 0) Then
            Return Nothing
        Else

            Return String.Format(CultureInfo.InvariantCulture,
                "DistributionTypeID IN ({0})", CreateIntegerFilterSpec(ids))

        End If

    End Function

    Private Shared Function CreateUserDistributionTypeFilter(ByVal project As Project) As String

        Dim ids As List(Of Integer) = GetInternalDistributionTypeIds(project)

        If (ids.Count = 0) Then
            Return Nothing
        Else

            Return String.Format(CultureInfo.InvariantCulture,
                "DistributionTypeID NOT IN ({0})", CreateIntegerFilterSpec(ids))

        End If

    End Function

    Private Shared Function GetInternalDistributionTypeIds(ByVal project As Project) As List(Of Integer)

        Dim ids As New List(Of Integer)
        Dim ds As DataSheet = project.GetDataSheet(DISTRIBUTION_TYPE_DATASHEET_NAME)

        For Each dr As DataRow In ds.GetData.Rows

            If (dr.RowState <> DataRowState.Deleted) Then

                If (dr(DISTRIBUTION_TYPE_IS_INTERNAL_COLUMN_NAME) IsNot DBNull.Value) Then

                    If (CInt(dr(DISTRIBUTION_TYPE_IS_INTERNAL_COLUMN_NAME)) = CInt(True)) Then
                        ids.Add(CInt(dr(ds.PrimaryKeyColumn.Name)))
                    End If

                End If

            End If

        Next

        Return ids

    End Function

    Private Shared Function CreateIntegerFilterSpec(ByVal ids As List(Of Integer)) As String

        Dim sb As New StringBuilder()

        For Each i As Integer In ids

            sb.Append(i.ToString(CultureInfo.InvariantCulture))
            sb.Append(",")

        Next

        Return sb.ToString.TrimEnd(CChar(","))

    End Function

End Class

File: DataFeedViews/DistributionValueDataFeedView.cs
// A SyncroSim Package for developing state-and-transition simulation models using ST-Sim.
// Copyright © 2007-2018 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.

using System;
using System.Text;
using System.Windows.Forms;
using System.Globalization;
using SyncroSim.Core;
using SyncroSim.Core.Forms;
using System.Diagnostics;
using System.Data;
using System.Collections.Generic;

namespace SyncroSim.STSim
{
    internal partial class DistributionValueDataFeedView
    {
        public DistributionValueDataFeedView()
        {
            InitializeComponent();
        }

        private BaseDataGridView m_Grid;
        private int? m_TypeId;

        protected override void InitializeView()
        {
            base.InitializeView();

            DataFeedView v = this.Session.CreateMultiRowDataFeedView(this.Scenario, this.ControllingScenario);
            this.PanelMain.Controls.Add(v);

            DataSheet ds = this.Project.GetDataSheet(Strings.DISTRIBUTION_TYPE_DATASHEET_NAME);

            try
            {
                this.m_TypeId = ds.ValidationTable.GetValue(Strings.DISTRIBUTION_TYPE_NAME_UNIFORM_INTEGER);
            }
            catch (Exception)
            {
                Debug.Assert(false);
            }
        }

        protected override void Dispose(bool disposing)
        {
            if (disposing && !this.IsDisposed)
            {
                if (this.m_Grid != null)
                {
                    this.m_Grid.CellBeginEdit -= OnGridCellBeginEdit;
                    this.m_Grid.CellEndEdit -= OnGridCellEndEdit;
                    this.m_Grid.CellFormatting -= OnGridCellFormatting;
                    this.m_Grid.CellValueChanged -= OnGridCellValueChanged;
                }

                if (components != null)
                {
                    components.Dispose();
                }
            }

            base.Dispose(disposing);
        }

        public override void LoadDataFeed(Core.DataFeed dataFeed)
        {
            base.LoadDataFeed(dataFeed);

            MultiRowDataFeedView v = (MultiRowDataFeedView)this.PanelMain.Controls[0];
            this.m_Grid = v.GridControl;
            v.LoadDataFeed(dataFeed, Strings.DISTRIBUTION_VALUE_DATASHEET_NAME);

            this.m_Grid.CellBeginEdit += OnGridCellBeginEdit;
            this.m_Grid.CellEndEdit += OnGridCellEndEdit;
            this.m_Grid.CellFormatting += OnGridCellFormatting;
            this.m_Grid.CellValueChanged += OnGridCellValueChanged;
        }

        public override void EnableView(bool enable)
        {
            if (this.PanelMain.Controls.Count > 0)
            {
                DataFeedView v = (DataFeedView)this.PanelMain.Controls[0];
                v.EnableView(enable);
            }
        }

        private void OnGridCellBeginEdit(object sender, DataGridViewCellCancelEventArgs e)
        {
            if (e.ColumnIndex == this.m_Grid.Columns[Strings.DISTRIBUTION_VALUE_DIST_TYPE_ID_COLUMN_NAME].Index)
            {
                DataGridViewRow dgv = this.m_Grid.Rows[e.RowIndex];
                DataSheet ds = this.Project.GetDataSheet(Strings.DISTRIBUTION_TYPE_DATASHEET_NAME);
                string filter = CreateUserDistributionTypeFilter(this.Project);
                DataView dv = new DataView(ds.GetData(), filter, ds.DisplayMember, DataViewRowState.CurrentRows);
                DataGridViewComboBoxCell Cell = (DataGridViewComboBoxCell)dgv.Cells[Strings.DISTRIBUTION_VALUE_DIST_TYPE_ID_COLUMN_NAME];

                Cell.DataSource = dv;
                Cell.ValueMember = ds.ValueMember;
                Cell.DisplayMember = ds.DisplayMember;
            }
            else if (e.ColumnIndex == this.m_Grid.Columns[Strings.DISTRIBUTION_VALUE_VALUE_DIST_TYPE_ID_COLUMN_NAME].Index)
            {
                DataGridViewRow dgv = this.m_Grid.Rows[e.RowIndex];
                DataSheet ds = this.Project.GetDataSheet(Strings.DISTRIBUTION_TYPE_DATASHEET_NAME);
                string filter = CreateKnownDistributionTypeFilter(this.Project);
                DataView dv = new DataView(ds.GetData(), filter, ds.DisplayMember, DataViewRowState.CurrentRows);
                DataGridViewComboBoxCell Cell = (DataGridViewComboBoxCell)dgv.Cells[Strings.DISTRIBUTION_VALUE_VALUE_DIST_TYPE_ID_COLUMN_NAME];

                Cell.DataSource = dv;
                Cell.ValueMember = ds.ValueMember;
                Cell.DisplayMember = ds.DisplayMember;
            }
        }

        private void OnGridCellEndEdit(object sender, DataGridViewCellEventArgs e)
        {
            if (e.ColumnIndex == this.m_Grid.Columns[Strings.DISTRIBUTION_VALUE_DIST_TYPE_ID_COLUMN_NAME].Index)
            {
                DataGridViewRow dgv = this.m_Grid.Rows[e.RowIndex];
                DataGridViewComboBoxCell Cell = (DataGridViewComboBoxCell)dgv.Cells[Strings.DISTRIBUTION_VALUE_DIST_TYPE_ID_COLUMN_NAME];
                DataGridViewComboBoxColumn Column = (DataGridViewComboBoxColumn)this.m_Grid.Columns[Strings.DISTRIBUTION_VALUE_DIST_TYPE_ID_COLUMN_NAME];

                Cell.DataSource = Column.DataSource;
                Cell.ValueMember = Column.ValueMember;
                Cell.DisplayMember = Column.DisplayMember;
            }
            else if (e.ColumnIndex == this.m_Grid.Columns[Strings.DISTRIBUTION_VALUE_VALUE_DIST_TYPE_ID_COLUMN_NAME].Index)
            {
                DataGridViewRow dgv = this.m_Grid.Rows[e.RowIndex];
                DataGridViewComboBoxCell Cell = (DataGridViewComboBoxCell)dgv.Cells[Strings.DISTRIBUTION_VALUE_DIST_TYPE_ID_COLUMN_NAME];
                DataGridViewComboBoxColumn Column = (DataGridViewComboBoxColumn)this.m_Grid.Columns[Strings.DISTRIBUTION_VALUE_VALUE_DIST_TYPE_ID_COLUMN_NAME];

                Cell.DataSource = Column.DataSource;
                Cell.ValueMember = Column.ValueMember;
                Cell.DisplayMember = Column.DisplayMember;
            }
        }

        private void OnGridCellFormatting(object sender, DataGridViewCellFormattingEventArgs e)
        {
            if (e.RowIndex < 0)
            {
                return;
            }

            //Is it an index we want?

            if (e.ColumnIndex != this.m_Grid.Columns[Strings.DISTRIBUTION_VALUE_VALUE_DIST_MIN_COLUMN_NAME].Index && 
                e.ColumnIndex != this.m_Grid.Columns[Strings.DISTRIBUTION_VALUE_VALUE_DIST_MAX_COLUMN_NAME].Index)
            {
                return;
            }

            //Set the default
            e.CellStyle.Format = "0.0000";

            //Does the cell have a value?

            DataGridViewCell TargetCell = this.m_Grid.Rows[e.RowIndex].Cells[e.ColumnIndex];

            if (TargetCell.Value == null || TargetCell.Value == DBNull.Value)
            {
                return;
            }

            //Is the distribution type 'Uniform Integer'?

            DataGridViewCell TypeCell = this.m_Grid.Rows[e.RowIndex].Cells[Strings.DISTRIBUTION_VALUE_VALUE_DIST_TYPE_ID_COLUMN_NAME];

            if (TypeCell.Value == null || TypeCell.Value == DBNull.Value || (!this.m_TypeId.HasValue))
            {
                return;
            }

            int ValTypeId = Convert.ToInt32(TypeCell.Value, CultureInfo.InvariantCulture);

            if (ValTypeId != this.m_TypeId.Value)
            {
                return;
            }

            e.CellStyle.Format = "0";
        }

        private void OnGridCellValueChanged(object sender, DataGridViewCellEventArgs e)
        {
            if (e.ColumnIndex == this.m_Grid.Columns[Strings.DISTRIBUTION_VALUE_VALUE_DIST_TYPE_ID_COLUMN_NAME].Index)
            {
                this.m_Grid.InvalidateRow(e.RowIndex);
            }
        }


        private static string CreateKnownDistributionTypeFilter(Project project)
        {
            List<int> ids = GetInternalDistributionTypeIds(project);
            Debug.Assert(ids.Count > 0);

            if (ids.Count == 0)
            {
                return null;
            }
            else
            {
                return string.Format(CultureInfo.InvariantCulture, "DistributionTypeID IN ({0})", CreateIntegerFilterSpec(ids));
            }
        }

        private static string CreateUserDistributionTypeFilter(Project project)
        {
            List<int> ids = GetInternalDistributionTypeIds(project);

            if (ids.Count == 0)
            {
                return null;
            }
            else
            {
                return string.Format(CultureInfo.InvariantCulture, 
                    "DistributionTypeID NOT IN ({0})", 
                    CreateIntegerFilterSpec(ids));
            }
        }

        private static List<int> GetInternalDistributionTypeIds(Project project)
        {
            List<int> ids = new List<int>();
            DataSheet ds = project.GetDataSheet(Strings.DISTRIBUTION_TYPE_DATASHEET_NAME);

            foreach (DataRow dr in ds.GetData().Rows)
            {
                if (dr.RowState != DataRowState.Deleted)
                {
                    if (Booleans.BoolFromValue(dr[Strings.DISTRIBUTION_TYPE_IS_INTERNAL_COLUMN_NAME]))
                    {
                        ids.Add(Convert.ToInt32(dr[ds.PrimaryKeyColumn.Name], CultureInfo.InvariantCulture));
                    }
                }
            }

            return ids;
        }

        private static string CreateIntegerFilterSpec(List<int> ids)
        {
            StringBuilder sb = new StringBuilder();

            foreach (int i in ids)
            {
                sb.Append(i.ToString(CultureInfo.InvariantCulture));
                sb.Append(",");
            }

            return sb.ToString().TrimEnd(',');
        }
    }
}
